libs/vkd3d/memory.c

```diff
@@ -380,6 +394,197 @@ static HRESULT vkd3d_memory_allocation_init(struct vkd3d_memory_allocation *allo
     return S_OK;
 }
 
+static void vkd3d_memory_chunk_insert_range(struct vkd3d_memory_chunk *chunk,
+        size_t index, VkDeviceSize offset, VkDeviceSize length)
+{
+    if (!vkd3d_array_reserve((void**)&chunk->free_ranges, &chunk->free_ranges_size,
+            chunk->free_ranges_count + 1, sizeof(*chunk->free_ranges)))
+    {
+        ERR("Failed to insert free range.\n");
+        return;
+    }
+
+    memmove(&chunk->free_ranges[index + 1], &chunk->free_ranges[index],
+            sizeof(*chunk->free_ranges) * (chunk->free_ranges_count - index));
+
+    chunk->free_ranges[index].offset = offset;
+    chunk->free_ranges[index].length = length;
+    chunk->free_ranges_count++;
+}
+
+static void vkd3d_memory_chunk_remove_range(struct vkd3d_memory_chunk *chunk, size_t index)
+{
+    chunk->free_ranges_count--;
+
+    memmove(&chunk->free_ranges[index], &chunk->free_ranges[index + 1],
+            sizeof(*chunk->free_ranges) * (chunk->free_ranges_count - index));
+}
+
+static HRESULT vkd3d_memory_chunk_allocate_range(struct vkd3d_memory_chunk *chunk, const VkMemoryRequirements *memory_requirements,
+        struct vkd3d_memory_allocation *allocation)
+{
+    struct vkd3d_memory_free_range *pick_range;
+    VkDeviceSize l_length, r_length;
+    size_t i, pick_index;
+
+    if (!chunk->free_ranges_count)
+        return E_OUTOFMEMORY;
+
+    pick_index = chunk->free_ranges_count;
+    pick_range = NULL;
+
+    for (i = 0; i < chunk->free_ranges_count; i++)
+    {
+        struct vkd3d_memory_free_range *range = &chunk->free_ranges[i];
+
+        if (range->offset + range->length - align(range->offset, memory_requirements->alignment) < memory_requirements->size)
+            continue;
+
+        /* Exact fit leaving no gaps */
+        if (range->length == memory_requirements->size)
+        {
+            pick_index = i;
+            pick_range = range;
+            break;
+        }
+
+        /* Alignment is almost always going to be 64 KiB, so
+         * don't worry too much about misalignment gaps here */
+        if (!pick_range || range->length > pick_range->length)
```

Comment:
I guess the intent here is to sub-allocate from the current largest range?

Reply:
yes.